src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala

```diff
@@ -0,0 +1,130 @@
+import com.fasterxml.jackson.module.scala.JsonScalaEnumeration
+import com.typesafe.config.Config
+import slick.jdbc.PostgresProfile
+import slick.jdbc.PostgresProfile.api._
+import tech.cryptonomic.conseil.tezos.PlatformDiscoveryOperations.DataType.DataType
+import tech.cryptonomic.conseil.tezos.PlatformDiscoveryOperations.KeyType.KeyType
+import tech.cryptonomic.conseil.tezos.{TezosDatabaseOperations => TezosDb}
+import tech.cryptonomic.conseil.util.DatabaseUtil
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+
+
+object PlatformDiscoveryOperations {
+
+  lazy val dbHandle: PostgresProfile.backend.Database = DatabaseUtil.db
+
+  private val tables = List(Tables.Blocks, Tables.Accounts, Tables.OperationGroups, Tables.Operations, Tables.Fees)
+
+  def getNetworks(config: Config): List[Network] = {
+    config.getObject("platforms").asScala.flatMap {
+      case (platform, strippedConf) =>
+        strippedConf.atKey(platform).getObject(platform).asScala.map {
+          case (network, _) =>
+            Network(network, network.capitalize, platform, network)
+        }.toList
+    }.toList
+  }
+
+  def getEntities(network: String)(implicit ec: ExecutionContext): Future[List[Entity]] = {
+    ApiOperations.countAll.map { counts =>
+      tables.map(_.baseTableRow.tableName).flatMap { tableName =>
+        counts.get(tableName).map { tableCount =>
+          Entity(
+            name = tableName,
+            displayName = makeDisplayName(tableName),
+            count = tableCount,
+            network = network
+          )
+        }.toList
+      }
+    }
+  }
+
+  private def makeDisplayName(name: String): String = {
+    name.capitalize.replace("_", " ")
+  }
+
+  def tableAttributes(tableName: String)(implicit ec: ExecutionContext): Future[List[Attributes]] = {
+    val tablesMap = tables.map(table => table.baseTableRow.tableName -> table)
+
+    val res = tablesMap.find(_._1 == tableName).map {
+      case (_, table) =>
+        dbHandle.run {
+          val overallCount = TezosDb.count(table)
+          DBIO.sequence(
+            table.baseTableRow.create_*.map { col =>
+              for {
+                overallCnt <- overallCount
+                distinctCnt <-  TezosDb.countDistinct(table.baseTableRow.tableName, col.name)
+              } yield  (col, distinctCnt, overallCnt)
+            }
+          )
+        }.map { distinctCountList =>
+          distinctCountList.map {
+            case (col, distinctCount, overallCount) =>
+              Attributes(
+                name = col.name,
+                displayName = makeDisplayName(col.name),
+                dataType = mapType(col.tpe.toString),
+                cardinality = distinctCount,
+                keyType = if(distinctCount == overallCount) KeyType.UniqueKey else KeyType.NonKey,
+                entity = table.baseTableRow.tableName
+              )
+          }
+        }
+    }
+    futOpt(res).map(_.toList.flatten)
```

Comment:
I can suggest splitting the DBIO creation from execution with something of the form
```scala
val dbAction: Option[DBIO[...]] = tablesMap.collect {
  ...
  DBIO.sequence(...)
}
Future.traverse(dbActions.toList)(dbHandle.run)
  .map(_.toList.flatten) //<- if needed still?
```

This would also remove the need to define `futOpt`

Reply:
made it a little different but with similar idea

Reply:
awesome